compiler/src/dotty/tools/dotc/macros/Typer.scala

```diff
@@ -0,0 +1,95 @@
+package dotty.tools.dotc
+package macros
+
+import core._
+import ast._
+import Trees._
+import Contexts.Context
+import Types._
+import StdNames._
+import Decorators._
+import Constants._
+import Symbols._
+import Flags._
+
+/** Typer for typing meta blocks
+ *
+ *    http://docs.scala-lang.org/sips/pending/inline-meta.html
+ *
+ *  Inside a meta block, e.g.
+ *
+ *    inline def is[T](a: Any): Any = meta {
+ *      q"$a.isInstanceOf[$T]"
+ *    }
+ *
+ *  The variables will get following types:
+ *
+ *    T : scala.meta.Type
+ *    a : scala.meta.Term
+ *
+ *  If it's an annotation macro, e.g.
+ *
+ *    class main extends StaticAnnotation {
+ *      inline def apply(defn: Any): Any = meta { ... }
+ *    }
+ *
+ *  The parameter `defn` gets the type `scala.meta.Tree` inside the meta block.
+ *
+ */
+class Typer extends typer.Typer {
+  import tpd._
+
+  override def typedThis(tree: untpd.This)(implicit ctx: Context): Tree = {
+    val tpdTree = super.typedThis(tree)
+
+    // transform type of identifier
+    val metaTypeRef = ctx.requiredClassRef("scala.meta.Term")
+
+    tpdTree.tpe match {
+      case tp: ThisType =>
+        tpdTree.withType(metaTypeRef)
+      case _ =>
+        tpdTree
+    }
+  }
+
+  override def typedIdent(tree: untpd.Ident, pt: Type)(implicit ctx: Context): Tree = {
+    val tpdTree = tryEither {
+      implicit ctx => super.typedIdent(tree, pt)
+    } { (failedVal, failedState) =>
+      tryEither {
+        implicit ctx => {
+          val name = if (tree.name.isTermName) tree.name.toTypeName else tree.name
+          super.typedIdent(untpd.cpy.Ident(tree)(name = name), pt)
+        }
+      } { (_, _) =>
+        failedState.commit()
+        failedVal
+      }
+    }
```

Comment:
Am I right in thinking that this logic makes sure that, inside meta blocks, we can refer to type parameters via term names?